### PR TITLE
Ports Z-Mimic

### LINKED ISSUE #8480: Port z-mimic and fix bee specific issues that will pop up
https://github.com/Baystation12/Baystation12/tree/dev/code/modules/multiz/zmimic

I have heard very good things about z-mimic and issues from multiple people regarding plane cubes, its maptick and performance with lower end clients.
Z-mimic is absurdly simple and I have heard good things from multiple-people.

### PR DESCRIPTION
**Pull request by itsmeow**
## About The Pull Request

Closes #8480

Ports a modified version of the port to Deadalus:
- https://github.com/DaedalusDock/daedalusdock/pull/280

Includes additional changes from:
- https://github.com/Citadel-Station-13/Citadel-Station-13-RP/pull/4979

As well as various hooks added for our codebase:
- mob/update_transform
- atom/Initialize
- mob/do_jump_animation forwarding
- update_icon() **this one is important**, most things that override without calling parent will not update properly on Z-mimics
- Area lighting support

atom/setDir now returns if the dir changed, so the movable z-mimic hook can update accordingly.

Added `ZMM_MANGLE_PLANES` to all objects that have emissives currently. AUTOMANGLE should cover the rest.

Tweaked lighting to use our current system

**new mutable_appearance(appearance) no longer defaults to FLOAT_PLANE**, change from this PR:

- https://github.com/tgstation/tgstation/pull/69838

This is important for Z-Mimic, since it creates mutable_appearances using other appearances and this shouldn't be set anyway on ones that copying existing appearances.

### Limitations

- No overlay lighting will render between zs
- Zstacks must be linear in zs, virtual zs are unsupported
- Performance tradeoff for anything with custom planes or emissive blockers (literally most /obj/item have emissive blockers)
- No ambient occlusion between zs at the moment
- Ghost eyes won't allow you to see between totally dark zs
- Things bigger than 1 turf may render oddly, singularities being a main culprit, especially since their space layering seems broken
- **vis_contents do not render**. at all. This includes airlocks currently. Anything that uses vis_contents is entirely ignored.

Known bug with blackness (APC/air alarm is turf below):
image

### Advantages

- **performance**
- More simplistic in its basic idea (no plane cube)
- Fixes some weird layering stuff with old vis_contents multiz (no plane cube)
- Supports multiz lighting
- Maintained by other codebases
- Easy to hook into and add features to (ex: jump animations render properly between zs)
- Fucking performance
- Performance

image

## Why It's Good For The Game

Z-Mimic is vastly superior to our current vis_contents system performance wise, and makes large openspace maps more viable.

It is a lot more technically complicated however so there is a compromise here. Our old system is extremely simple.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

nested zmimic

image

MultiZ light bleeding

image

image

No more emissive leaking

image

Successive zstack darkening, no AO right now

image

</details>

## Changelog
:cl: itsmeow, Lohikar, Kapu1178, CrimsonShrike
refactor: Rewrote multiz rendering to use Z-Mimic. Should vastly reduce lag caused by multiz rendering. Some things may not render properly.
fix: Emissives will no longer leak onto the z-level above.
add: Added multiz lighting. Nonmovable lights will now bleed up and down.
/:cl:

### CODE REVIEW

**Comment by Lohikar** [comment]
As a heads up, it looks like you've got an old revision of Z-Mimic here (this copy has broken layering) -- the copy on CitadelRP is more up to date.

**Comment by itsmeow** [comment]
> As a heads up, it looks like you've got an old revision of Z-Mimic here (this copy has broken layering) -- the copy on CitadelRP is more up to date.

Yep, I was just trying to see if I could get the basics to work. The system is really great, thank you so much for making it!

**Comment by Lohikar** [comment]
> Yep, I was just trying to see if I could get the basics to work. The system is really great, thank you so much for making it!

Thanks -- though you may want to make sure that you've read the assumptions list in the docs, ZM currently can't handle things like nonlinear z-stacks.

**Comment by itsmeow** [comment]
Yep, we don't do anything with that. Another thing I could use some help on: my lighting is currently solid black on the zmimic. I modified the lighting code since we don't use underlays and disabling the plane masters I wrote does not help. Not great with lighting personally, any ideas?

**Comment by Lohikar** [comment]
You probably want to reference the CitRP or Nebula versions for lighting-specific code if you're using object lighting, the daedalus copy has been modified for underlay lighting.

**Comment by itsmeow** [comment]
> You probably want to reference the CitRP or Nebula versions for lighting-specific code if you're using object lighting, the daedalus copy has been modified for underlay lighting.

That's exactly it, it looks like I wasn't assigning the appearance from the lighting object in copy_lighting_object.

Now my new biggest issue is AO, since our AO is done on the plane master level using a filter on the game plane.

File: code/modules/lighting/lighting_corner.dm [hunk 1]
```diff
@@ -139,34 +191,15 @@
...
-		master_SW.lighting_corner_NE = null
-		master_SW.lighting_corners_initialised = FALSE
-	if (master_NW)
-		master_NW.lighting_corner_SE = null
-		master_NW.lighting_corners_initialised = FALSE
-	if (needs_update)
-		SSlighting.corners_queue -= src
+	SHOULD_CALL_PARENT(FALSE)
+	stack_trace("QDEL CALLED ON LIGHTING CORNER. DO NOT DO THAT. PLEASE. \
+	THERE WAS A VERY RUDE AND NOT COOL COMMENT HERE TELLING YOU BAD THINGS, BUT I DON'T WANT TO PORT IT")
```
Comment (Kapu1178) [inline code/modules/lighting/lighting_corner.dm:201]: the PJB comment is soul it must be maintained for our children
```suggestion
	stack_trace("Ok, Look, TG, I need you to find whatever fucker decided to call qdel on a fucking lighting corner, \
	then tell him very nicely and politely that he is 100% retarded and needs his head checked. Thanks. Send them my regards by the way.")
```

Comment (itsmeow) [reply code/modules/lighting/lighting_corner.dm:201]: No!

**Comment by PowerfulBacon** [comment]
You mention performance a lot with no evidence or explanation about the performance. Of course we will put it on the server to see how it fares when compared to the crappy vis_contents system, but you shouldn't make claims about performance when there is no data to support / work against it. Especially since z-mimic should (in theory) have a higher CPU usage due to more work with the benefit being better overall maptick and client performance.

File: code/modules/lighting/lighting_corner.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/modules/lighting/lighting_corner.dm:201]: Might as well just change it to something more informative, like please let the coders know

Comment (itsmeow) [reply code/modules/lighting/lighting_corner.dm:201]: It's a runtime message. The coders know

**Comment by Crossedfall** [comment]
> The CPU usage is pretty minimal aside from initialization.

Show the data

**Comment by itsmeow** [comment]
That is my assumption, I don't have the tools to prove that.

**Comment by itsmeow** [comment]
Examining something worn by the player says it is on a different z level

**Comment by itsmeow** [comment]
TM went smoothly on Dead Station

**Review by PowerfulBacon — APPROVED** [review]
I'm a little worried about init time and runtime performance of extremely hot paths firing a lot for non-multi z maps but this seems fine

**Pull request merged**